File: app/src/main/java/com/battlelancer/seriesguide/ui/streams/TraktEpisodeHistoryLoader.java
package com.battlelancer.seriesguide.ui.streams;

import android.app.Activity;
import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.SgApp;
import com.battlelancer.seriesguide.traktapi.SgTrakt;
import com.battlelancer.seriesguide.traktapi.TraktCredentials;
import com.battlelancer.seriesguide.util.Errors;
import com.uwetrottmann.androidutils.AndroidUtils;
import com.uwetrottmann.androidutils.GenericSimpleLoader;
import com.uwetrottmann.trakt5.entities.HistoryEntry;
import com.uwetrottmann.trakt5.entities.UserSlug;
import com.uwetrottmann.trakt5.enums.HistoryType;
import com.uwetrottmann.trakt5.services.Users;
import java.util.List;
import retrofit2.Call;
import retrofit2.Response;

/**
 * Loads the last few episodes watched on trakt.
 */
class TraktEpisodeHistoryLoader extends GenericSimpleLoader<TraktEpisodeHistoryLoader.Result> {

    protected static final int MAX_HISTORY_SIZE = 50;

    static class Result {
        public List<HistoryEntry> results;
        public String emptyText;

        public Result(List<HistoryEntry> results, String emptyText) {
            this.results = results;
            this.emptyText = emptyText;
        }
    }

    TraktEpisodeHistoryLoader(Activity activity) {
        super(activity);
    }

    @Override
    public Result loadInBackground() {
        if (!TraktCredentials.get(getContext()).hasCredentials()) {
            return buildResultFailure(R.string.trakt_error_credentials);
        }

        List<HistoryEntry> history = null;
        try {
            Response<List<HistoryEntry>> response = buildCall().execute();
            if (response.isSuccessful()) {
                history = response.body();
            } else {
                if (SgTrakt.isUnauthorized(getContext(), response)) {
                    return buildResultFailure(R.string.trakt_error_credentials);
                }
                Errors.logAndReport(getAction(), response);
            }
        } catch (Exception e) {
            Errors.logAndReport(getAction(), e);
            return AndroidUtils.isNetworkConnected(getContext())
                    ? buildResultFailure() : buildResultFailure(R.string.offline);
        }

        if (history == null) {
            return buildResultFailure();
        } else {
            return new Result(history, getContext().getString(getEmptyText()));
        }
    }

    @NonNull
    protected String getAction() {
        return "get user episode history";
    }

    @StringRes
    protected int getEmptyText() {
        return R.string.user_stream_empty;
    }

    protected Call<List<HistoryEntry>> buildCall() {
        Users traktUsers = SgApp.getServicesComponent(getContext()).traktUsers();
        return traktUsers.history(UserSlug.ME, HistoryType.EPISODES, 1, MAX_HISTORY_SIZE,
                null, null, null);
    }

    private Result buildResultFailure() {
        return new Result(null, getContext().getString(R.string.api_error_generic,
                getContext().getString(R.string.trakt)));
    }

    private Result buildResultFailure(@StringRes int emptyTextResId) {
        return new Result(null, getContext().getString(emptyTextResId));
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/ui/streams/TraktEpisodeHistoryLoader.kt
package com.battlelancer.seriesguide.ui.streams

import android.content.Context
import androidx.annotation.StringRes
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.SgApp
import com.battlelancer.seriesguide.traktapi.SgTrakt
import com.battlelancer.seriesguide.traktapi.TraktCredentials
import com.battlelancer.seriesguide.util.Errors
import com.uwetrottmann.androidutils.AndroidUtils
import com.uwetrottmann.androidutils.GenericSimpleLoader
import com.uwetrottmann.trakt5.entities.HistoryEntry
import com.uwetrottmann.trakt5.entities.UserSlug
import com.uwetrottmann.trakt5.enums.HistoryType
import retrofit2.Call
import java.util.Calendar

/**
 * Loads the last few episodes watched on Trakt.
 */
open class TraktEpisodeHistoryLoader(context: Context) :
    GenericSimpleLoader<TraktEpisodeHistoryLoader.Result>(context) {

    data class HistoryItem(val headerTime: Long, val historyEntry: HistoryEntry)

    data class Result(
        val results: List<HistoryItem> = emptyList(),
        val emptyText: String
    )

    override fun loadInBackground(): Result {
        if (!TraktCredentials.get(context).hasCredentials()) {
            return buildResultFailure(R.string.trakt_error_credentials)
        }

        var history: List<HistoryEntry?>? = null
        try {
            val response = buildCall().execute()
            if (response.isSuccessful) {
                history = response.body()
            } else {
                if (SgTrakt.isUnauthorized(context, response)) {
                    return buildResultFailure(R.string.trakt_error_credentials)
                }
                Errors.logAndReport(action, response)
            }
        } catch (e: Exception) {
            Errors.logAndReport(action, e)
            return if (AndroidUtils.isNetworkConnected(context)) {
                buildResultFailure()
            } else {
                buildResultFailure(R.string.offline)
            }
        }

        return if (history == null) {
            buildResultFailure()
        } else {
            val calendar = Calendar.getInstance()
            Result(
                history.mapNotNull { entry ->
                    entry?.watched_at?.let {
                        HistoryItem(
                            calculateHeaderTime(calendar, it.toInstant().toEpochMilli()),
                            entry
                        )
                    }
                },
                context.getString(emptyText)
            )
        }
    }

    open val action: String
        get() = "get user episode history"

    @get:StringRes
    open val emptyText: Int
        get() = R.string.now_empty

    protected open fun buildCall(): Call<List<HistoryEntry?>?> {
        val traktUsers = SgApp.getServicesComponent(context).traktUsers()!!
        return traktUsers.history(
            UserSlug.ME,
            HistoryType.EPISODES,
            1,
            MAX_HISTORY_SIZE,
            null,
            null,
            null
        )
    }

    /**
     * Takes the action timestamp and "rounds" it down to shortly after midnight in the current
     * device time zone.
     */
    private fun calculateHeaderTime(calendar: Calendar, timestampMs: Long): Long {
        calendar.timeInMillis = timestampMs
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        calendar.set(Calendar.SECOND, 0)
        calendar.set(Calendar.MILLISECOND, 1)
        return calendar.timeInMillis
    }

    private fun buildResultFailure(): Result {
        return Result(
            emptyText = context.getString(
                R.string.api_error_generic,
                context.getString(R.string.trakt)
            )
        )
    }

    private fun buildResultFailure(@StringRes emptyTextResId: Int): Result {
        return Result(
            emptyText = context.getString(emptyTextResId)
        )
    }

    companion object {
        const val MAX_HISTORY_SIZE = 50
    }
}